org.eclipse.scanning.example/src/org/eclipse/scanning/example/malcolm/DummyMalcolmRecord.java

```diff
@@ -35,419 +35,308 @@
-        private final Status statusOk = StatusFactory.
-                getStatusCreate().getStatusOK();
-        private String methodName = "";
-
-        RPCServiceAsyncImpl(DummyMalcolmRecord record, String methodName) {
-            pvRecord = record;
-            this.methodName = methodName;
-        }
-
-        @Override
-		public void request(PVStructure args, RPCResponseCallback callback)
-        {
-		System.out.println("Got Async Request:");
-		System.out.println(args.toString());
-		receivedRPCCalls.put(methodName, args);
-
-            boolean haveControl = pvRecord.takeControl();
-            if (!haveControl)
-            {
-                handleError("Device busy", callback, haveControl);
-                return;
-            }
-
-
-            Structure mapStructure = FIELDCREATE.createFieldBuilder().
-				setId(CORE_ID+"Map:1.0").
-				createStructure();
-            PVStructure returnPvStructure = PVDATACREATE.createPVStructure(mapStructure);
-
-            if ("validate".equals(methodName)) {
-		returnPvStructure = args;
-            } else if ("configure".equals(methodName)) {
-		pvRecord.getPVStructure().getSubField(PVString.class, STATEVALUE).put("CONFIGURING");
-            } else if ("run".equals(methodName)) {
-                pvRecord.getPVStructure().getSubField(PVString.class, STATEVALUE).put("RUNNING");
-                try {
-				Thread.sleep(2000);
-			} catch (InterruptedException e1) {
-				e1.printStackTrace();
+		if (!underControl) {
+			underControl = true;
+			return true;
+		}
+		return false;
+	}
+
+	synchronized void releaseControl() {
+		underControl = false;
+	}
+
+	private class RPCServiceAsyncImpl implements RPCServiceAsync {
+
+		private DummyMalcolmRecord pvRecord;
+		private final Status statusOk = StatusFactory.getStatusCreate().getStatusOK();
+		private String methodName = "";
+
+		RPCServiceAsyncImpl(DummyMalcolmRecord record, String methodName) {
+			pvRecord = record;
+			this.methodName = methodName;
+		}
+
+		@Override
+		public void request(PVStructure args, RPCResponseCallback callback) {
+			System.out.println("Got Async Request:");
+			System.out.println(args.toString());
+			receivedRPCCalls.put(methodName, args);
+
+			boolean haveControl = pvRecord.takeControl();
+			if (!haveControl) {
+				handleError("Device busy", callback, haveControl);
+				return;
+			}
+
+			Structure mapStructure = FIELDCREATE.createFieldBuilder().setId(CORE_ID + "Map:1.0").createStructure();
+			PVStructure returnPvStructure = PVDATACREATE.createPVStructure(mapStructure);
+
+			if ("validate".equals(methodName)) {
+				returnPvStructure = args;
+			} else if ("configure".equals(methodName)) {
+				pvRecord.getPVStructure().getSubField(PVString.class, STATEVALUE).put("CONFIGURING");
+			} else if ("run".equals(methodName)) {
+				pvRecord.getPVStructure().getSubField(PVString.class, STATEVALUE).put("RUNNING");
+				try {
+					Thread.sleep(2000);
+				} catch (InterruptedException e1) {
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either re-interrupt this method or rethrow the "InterruptedException". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2142)